crates/services/consensus_module/poa/src/service.rs

```diff
@@ -199,13 +203,17 @@ where
             predefined_blocks,
             trigger,
             sync_task_handle,
+            clock,
         }
     }
 
-    fn extract_block_info(last_block: &BlockHeader) -> (BlockHeight, Tai64, Instant) {
+    fn extract_block_info(
+        clock: &C,
+        last_block: &BlockHeader,
+    ) -> (BlockHeight, Tai64, Instant) {
         let last_timestamp = last_block.time();
         let duration_since_last_block =
-            Duration::from_secs(Tai64::now().0.saturating_sub(last_timestamp.0));
+            Duration::from_secs(clock.now().0.saturating_sub(last_timestamp.0));
```

Comment:
> thoughts on saturating_sub vs checked_sub ? I think we should error out instead of saturating the result since it might be erroneous.

In general I agree with you. In this case I'm leaning towards thinking `checked_sub` is acceptable here because:

1. This is only relevant if the last block happened in the future, which is a pretty unlikely scenario.
2. The current behavior here is to tolerate previous blocks in the future, but "pretending" that they happened right now. Changing this breaks existing functionality.

I do think there is a lot of room for improvement in how the current code is structured, including this point. I'd be happy to create a follow-up ticket to address this if you think it's worth looking into.

Reply:
> additionally, we should try to avoid using inner fields explicitly for methods coming from trait impls. maybe you can impl CheckedSub for GetTime,.

Unfortunately, the orphan rule prevents me from implementing `CheckedSub` for `Tai64` (which GetTime::now outputs). The alternative would be to use a wrapper type which feels overkill in the current context. Especially since this is equivalent to the previous behavior.

Although I agree that using a wrapper type and implementing `CheckedSub` would be an improvement to the code.

Reply:
agree, a follow up issue will suffice here :)

Reply:
Follow-up issue created: https://github.com/FuelLabs/fuel-core/issues/2146